File: Rosu/programmazione/lezioni/java/huffman/src/ro/razvan/uniud/huffman/java/Node.java
package ro.razvan.uniud.huffman.java;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

class Node implements Comparable<Node> {

    private final char character;
    private final int weight;

    @Nullable
    private final Node left;
    @Nullable
    private final Node right;

    Node(final char character, final int weight) {
        this.character = character;
        this.weight = weight;
        left = null;
        this.right = null;
    }

    Node(@NotNull final Node left, @Nullable final Node right) {
        this.left = left;
        this.right = right;
        weight = left.weight + (right != null ? right.weight : 0);
        character = (char) 0;
    }

    char getCharacter() {
        return character;
    }

    int getWeight() {
        return weight;
    }

    @Nullable
    Node getLeft() {
        return left;
    }

    @Nullable
    Node getRight() {
        return right;
    }

    /**
     * @return se il nodo dell'albero di Huffman è una foglia
     */
    boolean isLeaf() {
        return left == null;
    }

    @Override
    public int compareTo(@NotNull Node o) {
        return Integer.compare(weight, o.weight);
    }

    @Override
    public String toString() {
        if (isLeaf()) {
            if (character == '@' || character == '\\') {
                return "\\" + character + ":" + weight;
            } else {
                return character + ":" + weight;
            }
        } else {
            return "@:" + weight;
        }
    }

}

File: Rosu/programmazione/laboratorio/java/progetti/huffman/src/ro/razvan/uniud/huffman/kotlin/Node.kt
package ro.razvan.uniud.huffman.kotlin

class Node : Comparable<Node> {

    private val ch: Char
    val weight: Int

    val left: Node?
    val right: Node?

    constructor(ch: Char, weight: Int = 0) {
        this.ch = ch
        this.weight = weight
        left = null
        right = null
    }

    constructor(left: Node, right: Node) {
        this.left = left
        this.right = right
        ch = 0.toChar()
        weight = left.weight + right.weight
    }

    /**
     * Ritorna se il nodo è una foglia dell'albero di Huffman
     */
    val isLeaf: Boolean
        get() = left == null

    val character: Char
        get() = ch

    override fun compareTo(other: Node): Int {
        return when {
            other.weight == weight -> 0
            other.weight > weight -> -1
            else -> 1
        }
    }

    /**
     * @ e \ sono caratteri speciali a cui va messo un \ in testa per poter essere riconosciuti
     * @return la stringa contenente il peso del carattere
     */
    override fun toString(): String = if (isLeaf) {

        if (ch == '@' || character == '\\') {
            "\\$character:$weight"
        } else {
            "$character:$weight"
        }

    } else {
        "@:$weight"
    }

}